lib/vector-buffers/src/config.rs

```diff
@@ -186,11 +244,18 @@ pub enum BufferType {
     /// This is more performant, but less durable. Data will be lost if Vector is restarted
     /// forcefully or crashes.
     #[configurable(title = "Events are buffered in memory.")]
-    #[serde(rename = "memory")]
+    #[serde(rename = "memory", serialize_with = "serialize_memory_config")]
     Memory {
+        // Two size options instead of a variant had been chosen to denote respective sizes so
+        // that the generated documentation will output what the parser will expect, a flat
+        // non-nested layout where exactly one of the two must be provided or else a default for
+        // max_events will be chosen.
         /// The maximum number of events allowed in the buffer.
-        #[serde(default = "memory_buffer_default_max_events")]
-        max_events: NonZeroUsize,
+        max_events: Option<NonZeroUsize>,
+
+        /// The maximum size across all events allowed in the buffer.
+        #[configurable(metadata(docs::type_unit = "bytes"))]
+        max_size: Option<NonZeroUsize>,
```

Comment:
I thought you could still accomplish the flattening with an enum:
```
        #[serde(default, flatten)]
        size: EventsOrSize,

#[serde(rename_all = "snake_case")]
enum EventsOrSize {
    MaxEvents(NonZeroUsize),
    MaxSize(NonZeroUsize),
}
```

Oh look, that's actually `MemoryBufferSize` which just needs the serde annotations.

Reply:
I gave that a shot, it was not generating the documentation as intended though